src/smt/smt_driver.h

```diff
@@ -0,0 +1,106 @@
+/******************************************************************************
+ * Top contributors (to current version):
+ *   Andrew Reynolds
+ *
+ * This file is part of the cvc5 project.
+ *
+ * Copyright (c) 2009-2022 by the authors listed in the file AUTHORS
+ * in the top-level source directory and their institutional affiliations.
+ * All rights reserved.  See the file COPYING in the top-level source
+ * directory for licensing information.
+ * ****************************************************************************
+ *
+ * The solver for SMT queries in an SolverEngine.
+ */
+
+#include "cvc5_private.h"
+
+#ifndef CVC5__SMT__SMT_DRIVER_H
+#define CVC5__SMT__SMT_DRIVER_H
+
+#include <vector>
+
+#include "expr/node.h"
+#include "smt/assertions.h"
+#include "smt/env_obj.h"
+#include "util/result.h"
+
+namespace cvc5::internal {
+namespace smt {
+
+class SmtSolver;
+class ContextManager;
+
+/**
+ * SMT driver class.
+ *
+ * The purpose of this class is to define algorithms for checking
+ * satisfiability beyond a single call to the underlying SMT solver. The
+ * default implementation, SmtDriverSingleCall, is used for invoking a
+ * single call to the SMT solver only.
+ */
+class SmtDriver : protected EnvObj
+{
+ public:
+  SmtDriver(Env& env, SmtSolver& smt, ContextManager* ctx);
+
+  /**
+   * Check satisfiability. This invokes the algorithm given by this driver
+   * for checking satisfiability.
+   *
+   * @param assumptions The assumptions for this check-sat call, which are
+   * temporary assertions.
+   */
+  Result checkSatisfiability(const std::vector<Node>& assumptions);
+
+ protected:
+  /**
+   * Check satisfiability next, return the result.
+   *
+   * If checkAgain is set to true, then this driver will be called to
+   * getNextAssertions as described below.
+   *
+   * If checkAgain is not set or set to false, then the returned result
```

Comment:
Is `checkAgain` going to be strictly treated as an output? The documentation is not very clear on this. If yes, why not return a pair of `Result` and `bool`?

Reply:
I prefer having the argument since it documents the meaning of the bool in its name.

I will update the documentation to make this clear.

Reply:
Hmm, I don't like this pattern. In the current implementation of `SmtDriverSingleCall::checkSatNext` the flag is not even used. The documentation argument is a bit moot here since who is the one that needs to check again? It's the caller of the method and not the method itself. So one has to read the documentation anyways to figure out who has to check again. Returning a `std::pair<Result, bool>` would be a lot clearer.

Reply:
If we are wrapping everything into the return value, I'd prefer just `Result`, which already can be modified to say that another check should be performed, i.e. `UnknownExplanation = REQUIRES_FULL_CHECK`